runtime/strings/strings.cpp

```diff
@@ -0,0 +1,291 @@
+	// search returns the end of the range if it is not found, but we want -1 in such a case.
+	return (ret < haystack->b.len)?ret:-1;
+}
+
+int64_t hook_STRING_rfind(const string * haystack, const string * needle, int64_t pos) {
+	// The semantics of rfind position are strange, it is the last position at which
+	// the match can _start_, which means the end of the haystack needs to be pos + len(needle),
+	// or the end of the haystack, if that's less.
+	pos += needle->b.len;
+	if (pos <= needle->b.len) {
+		return -1;
+	}
+	auto end = (pos < haystack->b.len)?pos:haystack->b.len;
+	auto out = std::find_end(&haystack->data[0], &haystack->data[end],
+                  &needle->data[0], &needle->data[needle->b.len]);
+	auto ret = &*out - &haystack->data[0];
+	return (ret < end)?ret:-1;
+}
+
+int64_t hook_STRING_findChar(const string * haystack, const string * needle, int64_t pos) {
+	if (needle->b.len > 1) return -1;
+	return hook_STRING_find(haystack, needle, pos);
+}
+
+int64_t hook_STRING_rfindChar(const string * haystack, const string * needle, int64_t pos) {
+	if (needle->b.len > 1) return -1;
+	return hook_STRING_rfind(haystack, needle, pos);
+}
+
+string * makeString(const KCHAR * input) {
+	auto len = strlen(input);
+	auto ret = static_cast<string *>(malloc(sizeof(string) + len));
+	for (unsigned i = 0; i < len; ++i) {
+		ret->data[i] = input[i];
+	}
+	ret->b.len = len;
+	return ret;
+}
+
+string * hook_STRING_float2string(const double input) {
+	std::stringstream out;
+	out << std::setprecision(14) << input;
+	return makeString(out.str().c_str());
+}
+
+double hook_STRING_string2float(const string * input) {
+	auto si = std::string(input->data, input->b.len);
+	return std::stod(si);
+}
+
+const string * hook_STRING_string2token(const string * input) {
+	return input;
+}
+
+// uses std::string for memory allocation, double copy here is not the best, but at least the
+// first copy is released at the end of the function.
+string * hook_STRING_replaceAll(const string * input, const string * matcher, const string * replacer) {
+	std::string ret;
+	auto rs = std::string(replacer->data, replacer->b.len);
+	for (auto i = 0; i < input->b.len;) {
+		if (input->data[i] == matcher->data[0]) {
+			auto found = false;
+			for (auto j = 0; j < matcher->b.len; ++j) {
+				if (i+j >= input->b.len) {
+					found = false;
+					break;
+				}
+				if  (input->data[i+j] != matcher->data[j]) {
+					found = false;
+					break;
+				}
+				found = true;
+			}
+			if (found) {
+				ret += rs;
+				i += matcher->b.len;
+			} else {
+				ret.push_back(input->data[i]);
+				++i;
+			}
+		} else {
+			ret.push_back(input->data[i]);
+			++i;
+		}
+	}
+	return makeString(ret.c_str());
+}
+
+// uses std::string for memory allocation, double copy here is not the best, but at least the
+// first copy is released at the end of the function.
+string * hook_STRING_replace(const string * input, const string * matcher, const string * replacer, int64_t occurences) {
```

Comment:
this should be the same as replaceAll except it replaces at most occurrences times.